product_restricted_type/models/product_template.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2018-2024 ForgeFlow S.L.
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    @api.onchange("categ_id")
+    def _onchange_categ_id(self):
+        if self.categ_id and self.categ_id.restricted_product_type:
+            self.type = self.categ_id.restricted_product_type
+
+    @api.onchange("type")
+    def _onchange_type(self):
+        if self.type:
+            return {
+                "domain": {
+                    "categ_id": [("restricted_product_type", "in", [self.type, False])]
+                }
+            }
+        else:
+            return {"domain": {"categ_id": []}}
```

Comment:
Returning a domain from an `onchange` method is deprecated.

Here, we'll need to create an auxiliar var called `allowed_categ_ids` which will contain the allowed Product Categories to be assigned on the Product to be then set in the domain like: `[('id', 'in', allowed_categ_ids)]`